server/test/socketManager-test.js

```diff
@@ -0,0 +1,517 @@
+        const idFromPath = socketManager.getIdFromPath(pathname);
+        assert(idFromPath === id, 'Did not correctly validate the pathname');
+      });
+
+      it('should throw for invalid paths', () => {
+        const getId = () => socketManager.getIdFromPath(id);
+        assert.throws(getId, Error, /\binvalid\b|\bdoes not exist\b/i);
+      });
+    });
+  });
+
+  describe('socket message handlers', function() {
+    let id, event, socket, responseEvent;
+
+    async function broadcastTest(needle, event, node, socket) {
+      // first confirming integrity of the test
+      assert.include(event, needle, `The event should contain ${needle}`);
+      assert(
+        !node.http.channel(needle),
+        `Should not have ${needle} channel before broadcast has been sent`
+      );
+      socket.fire('broadcast', event);
+      await sleep(200);
+      assert(
+        node.http.channel(needle),
+        `Node server doesn't have a channel for ${needle}`
+      );
+    }
+
+    beforeEach('setup client and some handlers', async function() {
+      id = 'socket.io';
+      event = 'watch chain';
+      subscribeEvent = 'block connect';
+      responseEvent = 'heard event';
+
+      // adding default node and wallet client for testing handlers
+      if (nclient.opened) await nclient.close();
+      if (wclient.opened) await wclient.close();
+      await socketManager.addClients(id, { node: nclient, wallet: wclient });
+      socket = bsock.connect(ports.manager);
+      socket.on('error', e => {
+        throw e;
+      });
+      await socket.call('auth', apiKey);
+    });
+
+    afterEach('cleanup sockets and clients', async function() {
+      await socketManager.removeClients(id);
+      if (socket.connected) await socket.destroy();
+    });
+
+    it('should send broadcast messages to the correct servers', async function() {
+      await broadcastTest('chain', event, node, socket);
+    });
+
+    it('should be able able to interact with other client types, like wallet', async function() {
+      const wdb = node.plugins.walletdb;
+      let sockets = wdb.http.io.sockets.values();
+      for (const sock of sockets) {
+        assert(
+          sock.channel(`w:${walletId}`),
+          `Socket should not already be subscribed to wallet "${walletId}"`
+        );
+      }
+      await socket.call('dispatch', 'wallet join', walletId);
+      let joined = false;
+      sockets = wdb.http.io.sockets.values();
+      for (const sock of sockets) {
+        if (sock.channel(`w:${walletId}`)) joined = true;
+      }
+      assert(joined, `Socket did not successfully dispatch join to wallet`);
+      await socket.call('dispatch', 'wallet leave', walletId);
+    });
+
+    it('should create new channels for new subscriptions', async function() {
+      const eventName = 'test event';
+      const responseEvent = 'heard event';
+      const subscription = `node-${id}:${eventName}-${responseEvent}`;
```

Comment:
It would be nice if the single source of truth function could be used here

Reply:
Totally right. I'll move that function out into the class and then I can access it here.